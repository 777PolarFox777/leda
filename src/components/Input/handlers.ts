import * as React from 'react';
import { isFunction } from 'lodash';
import { CustomEventHandler, SetState } from '../../commonTypes';
import { InputProps, InputState } from './types';
import { isSymbolAllowed, isSymbolForbidden, transformToCase } from './helpers';

export const createChangeHandler = (
  props: InputProps, state: InputState, setValue: SetState<string>,
): CustomEventHandler<React.ChangeEvent<HTMLInputElement>> => ev => {
  const {
    maxLength, forbiddenSymbols, allowedSymbols, letterCase, value: valueProp, name, onChange,
  } = props;

  const { target: { value } } = ev;

  if (maxLength && value.length > maxLength) return;

  if (isSymbolForbidden(value, forbiddenSymbols)) return;

  if (!isSymbolAllowed(value, allowedSymbols)) return;

  const newValue = letterCase ? transformToCase(value, letterCase) : value;

  if (valueProp === undefined) setValue(value);

  const customEvent = {
    ...ev,
    component: {
      value: newValue,
      name,
    },
  };

  if (isFunction(onChange)) onChange(customEvent);
};

export const createClearHandler = (
  props: InputProps, state: InputState, setValue: SetState<string>,
): CustomEventHandler<React.MouseEvent<HTMLInputElement>> => ev => {
  const {
    value: valueProp, name, onChange,
  } = props;

  ev.preventDefault();

  const customEvent = {
    ...ev,
    component: {
      value: '',
      name,
    },
  };

  if (valueProp === undefined) setValue('');

  if (isFunction(onChange)) onChange(customEvent);
};

export const createBlurHandler = (
  props: InputProps, state: InputState, setFocused: SetState<boolean>, validate: () => boolean,
): React.FocusEventHandler<HTMLInputElement> => ev => {
  const { onBlur, name } = props;

  setFocused(false);

  const newValid = validate();

  if (isFunction(onBlur)) {
    const customEvent = {
      ...ev,
      component: {
        value: ev.target.value,
        name,
        isValid: newValid,
      },
    };

    onBlur(customEvent);
  }
};

export const createFocusHandler = (
  props: InputProps, state: InputState, setFocused: SetState<boolean>,
): React.FocusEventHandler<HTMLInputElement> => ev => {
  const { name, onFocus } = props;

  const { isValid } = state;

  setFocused(true);

  const customEvent = {
    ...ev,
    component: {
      value: ev.target.value,
      name,
      isValid,
    },
  };

  if (isFunction(onFocus)) onFocus(customEvent);
};

export const createKeyDownHandler = (
  props: InputProps,
): React.KeyboardEventHandler<HTMLInputElement> => ev => {
  const { name, onEnterPress } = props;

  if (ev.key === 'Enter' && isFunction(onEnterPress)) {
    const event = {
      ...ev,
      component: {
        name,
        value: ev.currentTarget.value,
      },
    };
    onEnterPress(event);
  }
};
