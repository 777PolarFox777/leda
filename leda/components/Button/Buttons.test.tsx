// @ts-nocheck
import React from 'react';
import {
  render,
  screen,
} from '@testing-library/react';
import { Button } from './index';
import { Input } from '../Input';


describe('Check Button snapshots collection', () => {
  test('is Button changed?', () => {
    const wrapper = render((
      <Button>test</Button>
    ));
    expect(wrapper.container)
      .toMatchSnapshot();
  });
  test('is Button with isLoading attributes changed?', () => {
    const wrapper = render((
      <Button isLoading>test</Button>
    ));
    expect(wrapper.container)
      .toMatchSnapshot();
  });
  test('is Button with isDisabled attributes changed?', () => {
    const wrapper = render((
      <Button isDisabled>test</Button>
    ));
    expect(wrapper.container)
      .toMatchSnapshot();
  });
  test('is Button with Form and onClick props changed?', () => {
    const wrapper = render((
      <div>
        <Input form="test" name="in" isRequired />
        <Button onClick={jest.fn()} form="test" isDisabled>test</Button>
      </div>
    ));
    expect(wrapper.container)
      .toMatchSnapshot();
  });
});
describe('Check Button basic condition set', () => {
  test('is Button have Primary class?', () => {
    render((
      <Button _primary>test</Button>
    ));
    expect(document.querySelectorAll('button.primary'))
      .toHaveLength(1);
  });
  test('is Button have Secondary class?', () => {
    render((
      <Button _secondary>test</Button>
    ));
    expect(document.querySelectorAll('button.secondary'))
      .toHaveLength(1);
  });
  test('is Button have Success class?', () => {
    render((
      <Button _success>test</Button>
    ));
    expect(document.querySelectorAll('button.success'))
      .toHaveLength(1);
  });
  test('is Button have Warning class?', () => {
    render((
      <Button _warning>test</Button>
    ));
    expect(document.querySelectorAll('button.warning'))
      .toHaveLength(1);
  });
  test('is Button have Danger class?', () => {
    render((
      <Button _danger>test</Button>
    ));
    expect(document.querySelectorAll('button.danger'))
      .toHaveLength(1);
  });
});
describe('Check Button advanced condition set', () => {
  test('is Button have Loading class?', () => {
    render((
      <Button isLoading>test</Button>
    ));
    expect(document.querySelectorAll('button.loading'))
      .toHaveLength(1);
  });
  test('is Button with isLoading attributes don`t run onClick?', () => {
    const onClickHandler = jest.fn();
    render((
      <Button isLoading onClick={onClickHandler}>test</Button>
    ));
    expect(onClickHandler)
      .not
      .toHaveBeenCalled();
  });
  test('is Button have Disabled class?', () => {
    render((
      <Button isDisabled>test</Button>
    ));
    expect(document.querySelectorAll('button.disabled'))
      .toHaveLength(1);
  });
  test('is Button with isDisabled attributes don`t run onClick?', () => {
    const onClickHandler = jest.fn();
    render((
      <Button isDisabled onClick={onClickHandler}>test</Button>
    ));
    expect(onClickHandler)
      .not
      .toHaveBeenCalled();
  });
});
describe('Check Button size set', () => {
  test('is Button have Small size class?', () => {
    render((
      <Button _small>test</Button>
    ));
    expect(document.querySelectorAll('button.small'))
      .toHaveLength(1);
  });
  test('is Button have Large size class?', () => {
    render((
      <Button _large>test</Button>
    ));
    expect(document.querySelectorAll('button.large'))
      .toHaveLength(1);
  });
  test('is Button have Block class?', () => {
    render((
      <Button _block>test</Button>
    ));
    expect(document.querySelectorAll('button.block'))
      .toHaveLength(1);
  });
});
describe('Check Button Form validation role collection', () => {
  test('is Button click and assigned Form submit right?', () => {
    const onClick = jest.fn();

    render((
      <div>
        <Input form="test" name="email" />
        <Button onClick={onClick} form="test">test</Button>
      </div>
    ));
    screen.getByText('test').click();
    expect(onClick).toBeCalledTimes(1);
  });
  test('is Button click and all From field valid?', () => {
    const onSubmitValid = jest.fn();
    const onSubmitInvalid = jest.fn();
    const email = 'some@email.ru';

    render((
      <div>
        <Input form="test" name="email" invalidMessage="error" isRequired validator="email" value={email} />
        <Button onClick={onSubmitValid} onValidationFail={onSubmitInvalid} form="test">test</Button>
      </div>
    ));
    screen.getByText('test').click();
    expect(onSubmitValid).toBeCalledTimes(1);
    expect(onSubmitInvalid).toBeCalledTimes(0);
  });
  test('is Button click and some field is not valid?', () => {
    const onSubmitValid = jest.fn();
    const onSubmitInvalid = jest.fn();
    const email = '-----';

    render((
      <div>
        <Input form="test" name="email" invalidMessage="error" isRequired validator="email" value={email} />
        <Button onClick={onSubmitValid} onValidationFail={onSubmitInvalid} form="test">test</Button>
      </div>
    ));
    screen.getByText('test').click();
    expect(onSubmitValid).toBeCalledTimes(0);
    expect(onSubmitInvalid).toBeCalledTimes(1);
  });
});
