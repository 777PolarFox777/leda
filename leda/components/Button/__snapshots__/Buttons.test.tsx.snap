// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check Button snapshots collection is Button changed 1`] = `
<div>
  <button
    class="button-wrapper"
    type="button"
  >
    test
  </button>
</div>
`;

exports[`Check Button snapshots collection is Button changed? 1`] = `
<div>
  <button
    class="button-wrapper"
    type="button"
  >
    test
  </button>
</div>
`;

exports[`Check Button snapshots collection is Button with Form and onClick props changed 1`] = `
<div>
  <div>
    <div
      class="input-wrapper"
    >
      <div
        class="input-element-wrapper required"
      >
        <input
          aria-invalid="false"
          aria-required="true"
          class="input-element"
          form="test"
          name="in"
          value=""
        />
      </div>
    </div>
    <button
      class="button-wrapper disabled"
      type="button"
    >
      test
    </button>
  </div>
</div>
`;

exports[`Check Button snapshots collection is Button with Form and onClick props changed? 1`] = `
<div>
  <div>
    <div
      class="input-wrapper"
    >
      <div
        class="input-element-wrapper required"
      >
        <input
          aria-invalid="false"
          aria-required="true"
          class="input-element"
          form="test"
          name="in"
          value=""
        />
      </div>
    </div>
    <button
      class="button-wrapper disabled"
      type="button"
    >
      test
    </button>
  </div>
</div>
`;

exports[`Check Button snapshots collection is Button with isDisabled attributes changed? 1`] = `
<div>
  <button
    class="button-wrapper disabled"
    type="button"
  >
    test
  </button>
</div>
`;

exports[`Check Button snapshots collection is Button with isDisabled changed 1`] = `
<div>
  <button
    class="button-wrapper disabled"
    type="button"
  >
    test
  </button>
</div>
`;

exports[`Check Button snapshots collection is Button with isLoading attributes changed? 1`] = `
<div>
  <button
    class="button-wrapper loading"
    type="button"
  >
    test
  </button>
</div>
`;

exports[`Check Button snapshots collection is Button with isLoading changed 1`] = `
<div>
  <button
    class="button-wrapper loading"
    type="button"
  >
    test
  </button>
</div>
`;
